### PR TITLE
Richard/win condition

### PR DESCRIPTION
**Pull request by yds05238**
## Overview
Added win condition to the game. Game ends when planet reaches a specified mass units. Once  a game ends, start a countdown which triggers game reset (switch to loading page). Switched to using disposing and re-initializing the loading and game scenes in case of reset. 

## Changes Made

- Changed Game Update didPlayerWin method to check planet mass
- Defined planet mass needed to win and isWinner() method in Planet Model
- Handling winning in Game Scene update method (resetting the game after a countdown)
- Updated Game Scene's reset method to dispose resources/remove children
- Added reset method to Loading Scene 
- Changed Game App's update method to dispose all scenes then re-initialize them on reset

### CODE REVIEW

File: source/CIPlanetModel.h [hunk 1]
```diff
@@ -37,6 +40,11 @@ class PlanetModel {
     
     /** Scene graph node for the planet */
     std::shared_ptr<PlanetNode> _planetNode;
+    
+    /** Whether this planet satisfies winning conditions */
+    bool _winner;
+    /** Whether this planet satisfies losing conditions */
+    bool _loser;
```
[the remark this answers is not in the document]
Comment (yds05238) [reply source/CIPlanetModel.h:47]: Will be removing these 2 fields.

File: source/CIGameScene.cpp [hunk 2]
```diff
@@ -154,14 +161,30 @@ void GameScene::update(float timestep) {
...
+    } else if (_countdown == 0) {
+        // TODO: return to main menu instead of resetting the game
+        reset();
+        return;
+    }
+    
+    if (_planet->isWinner()) {
+        // handle winning
+        CULog("Game won.");
+        _countdown = 120;
```
[the remark this answers is not in the document]
Comment (yds05238) [reply source/CIGameScene.cpp:180]: Do you mean like create a separate timestep field to use or is there a pre-existing one that I can take advantage of?

File: source/CIGameScene.cpp [hunk 3]
```diff
@@ -140,6 +144,9 @@ void GameScene::reset() {
     _farSpace->setAngle(0.0f);
     
     _draggedStardust = NULL;
+    
+    removeAllChildren();
```
[the remark this answers is not in the document]
Comment (yds05238) [reply source/CIGameScene.cpp:148]: I can remove them but reset then just freezes the game

File: source/CILoadingScene.cpp [hunk 4]
```diff
@@ -92,6 +92,22 @@ void LoadingScene::dispose() {
     _progress = 0.0f;
 }
 
+/**
+ * Resets the resources allocated to this mode to default.
+ */
+void LoadingScene::reset() {
```
Comment (wlong123) [inline source/CILoadingScene.cpp:98]: is this reset method necessary?

File: source/CIGameUpdate.h [hunk 5]
```diff
@@ -140,10 +140,9 @@ class GameUpdate {
      * Returns whether the player sending the game update has won.
      */
     bool didPlayerWin() {
-        // TODO: need to implement win condition
-        return false;
+        // TODO: need to implement network win/lose condition
```
Comment (wlong123) [inline source/CIGameUpdate.h:143]: Can get rid of the to do here

File: source/CILoadingScene.cpp [hunk 6]
```diff
@@ -103,6 +102,18 @@ void LoadingScene::dispose() {
  * @param timestep  The amount of time (in seconds) since the last frame
  */
 void LoadingScene::update(float progress) {
+    if (isActive()) { // initially set all values to default
```
Comment (wlong123) [inline source/CILoadingScene.cpp:105]: So I just tested and the stuff in if (isActive()) is not necessary so please delete it

**Review by wlong123 — APPROVED** [review]

**Pull request merged**